draft-ietf-ppm-dap.md

```diff
@@ -640,6 +636,82 @@ DAP has three major interactions which need to be defined:
 Each of these interactions is defined in terms of "resources". In this section
 we define these resources and the messages used to act on them.
 
+## Message Encoding
+
+We use the presentation language defined in {{!RFC8446, Section 3}} to define
+messages in the DAP protocol, with the following deviations.
+
+{{Section 3.7 of !RFC8446}} defines a syntax for structure fields whose values
+are constants. In this document, we instead use that notation to describe
+specific variants of structures containing enumerated types, described in
+{{!RFC8446, Section 3.8}}.
+
+For example, suppose we have an enumeration and a structure defined as follows:
+
+~~~
+enum {
+  number(0),
+  string(1),
+  (255)
+} ExampleEnum;
+
+struct {
+  uint32 always_present;
+  ExampleEnum type;
+  select (ExampleStruct.type) {
+    case number: uint32 a_number;
+    case string: opaque a_string<0..10>;
+  };
+} ExampleStruct;
+~~~
+
+Then we describe the specific variant of `ExampleStruct` where `type == number`
+like so:
+
+~~~
+struct {
+  /* Field exists regardless of variant */
+  uint32 always_present;
+  ExampleEnum type = number;
+  /* Only fields included in the type == number variant is described */
+  uint32 a_number;
+} ExampleStruct;
```

Comment:
This semantic is defined in [variant](https://datatracker.ietf.org/doc/html/rfc8446#section-3.8) section, right? Are we making it `select` readable here? If so, do we need to do this for every structure? Or there is a special case where illustrating the above way makes it clear?

Seeing same type definition multiple times with different fields could confuse readers. Instead of extending the semantics of `struct`, should this type be called `const` or `type`?

Reply:
The pertinent examples are various cases of constructing `PrepareResp` as shown [here](https://datatracker.ietf.org/doc/html/draft-ietf-ppm-dap-11#section-4.5.1.2-9) and [here](https://datatracker.ietf.org/doc/html/draft-ietf-ppm-dap-11#section-4.5.1.2-15) or [here](https://datatracker.ietf.org/doc/html/draft-ietf-ppm-dap-11#section-4.5.2.2-13).

I like the idea of annotating such partial constants somehow to make it clear this special syntax is in use. `const` doesn't feel right because only part of the structure is constant. Maybe `struct specialized` to indicate we are describing a particular specialization of the structure:

```
struct specialized {
  ReportID report_id;
  PrepareRespState prepare_resp_state = continue;
  opaque payload<0..2^32-1> = outbound;
} PrepareResp;
```

Or perhaps `struct variant` is better, since we are dealing with enums here, and "specialized" implies generics.

Reply:
`struct variant` sounds good to me.

Reply:
I'm confused by the need for special syntax. What is a "partial constant"?

Reply:
[RFC 8446 3.7](https://datatracker.ietf.org/doc/html/rfc8446#section-3.7) defines constant fields of structs:

```
 struct {
          T1 f1 = 8;  /* T.f1 must always be 8 */
          T2 f2;
      } T;
```

This notation collides with our usage, where `PrepareResp.prepare_resp_state` can have many values, but we want to describe the variant where `prepare_resp_state = continue`.

Reply:
In the example given here, can't `f1` have many values?

Reply:
No, that is the point of the RFC 8446 constant notation. The comment says "T.f1 must always be 8". If you click through to 8446 s3.7 it says 'Fields and variables may be assigned a fixed value using "="'.

Reply:
Got it, thanks for clarifying. I think this could be clearer in the PR.